adversarial_workflow/utils/citations.py

```
@@ -0,0 +1,607 @@
+        return {}
+
+
+def save_cache(cache_path: Path, cache: dict[str, dict]) -> None:
+    """Save URL cache to disk."""
+    with open(cache_path, "w") as f:
+        json.dump(cache, f, indent=2)
+
+
+def get_cache_key(url: str) -> str:
+    """Generate a cache key for a URL."""
+    return hashlib.md5(url.encode()).hexdigest()
```

Ruff (0.14.14)

[error] Probable use of insecure hash functions in `hashlib`: `md5`

(S324)